### PR TITLE
feat(use-router): change default scroll restoration behavior

### LINKED ISSUE #791: Add scroll-related options to useRouter()
Hello,

I've noticed a bug-like behavior when handling state with query parameters. Every time the query changes, the scroll position resets to the top of the page. This can be disruptive to the user experience, especially on pages with a lot of content.
To address this, I suggest adding scroll-related options to the useRouter() hook. This would allow developers to control scroll behavior when navigating or updating query parameters.

If you're open to it, I'd be willing to work on this feature. Would you be interested in a pull request that implements this functionality?

Thank you for your consideration.

### PR DESCRIPTION
**Pull request by tylersayshi**
fixes #791

### Will always scroll

- router.reload
- router.back
- `<Link to={foo} />`

### Scrolls when path changes

- router.replace 
- router.push

### CODE REVIEW

**Comment by tylersayshi** [comment]
I am skeptical that `preserveScroll: true` is the best default for all scenarios, so I'd like to propose a couple scenarios:

1. going back to pages previously visited should restore scroll
2. clicking hash links should do a smooth scroll on that page to that link
3. updating query params `?foo=bar` probably should not trigger a scroll action.

Idea: 

maybe we consider different defaults in different scenarios:
- router.back & router.reload `preserveScroll: true`
- - router.replace & router.push `preserveScroll: false`
- `<Link to={path} />` scrolls only when hash is present in path by default (else does not scroll)

I think if we follow a more declarative ruleset along these lines we can remove the `waku_new_path` and also avoid devs being surprised by the scroll restoration in scenarios where they normally would not expect it to be restored.

File: packages/waku/src/router/client.ts [hunk 1]
```diff
@@ -299,18 +301,6 @@ class ErrorBoundary extends Component<
...
-// Eventually replaces Router and so on
-// -----------------------------------------------------
-
-type NewChangeRoute = (
-  route: RouteProps,
-  options?: {
-    skipRefetch?: boolean;
-  },
-) => void;
-
```
Comment (dai-shi) [inline packages/waku/src/router/client.ts:313]: Oops. Thanks.

File: packages/waku/src/router/client.ts [hunk 2]
```diff
@@ -344,7 +334,18 @@ const InnerRouter = ({
...
+  const handleScroll = useCallback(() => {
+    const { hash } = window.location;
+    const { state } = window.history;
+    const element = hash && document.getElementById(hash.slice(1));
+    window.scrollTo({
+      left: 0,
+      top: element ? element.getBoundingClientRect().top + window.scrollY : 0,
+      behavior: state?.waku_new_path ? 'instant' : 'auto',
+    });
+  }, []);
```
Comment (dai-shi) [inline packages/waku/src/router/client.ts:346]: Can we define this outside `useRouter`?

**Review by dai-shi — COMMENTED** [review]
looks good to me otherwise.

File: packages/waku/src/router/client.ts (hunk 2 above)
Comment (tylersayshi) [reply packages/waku/src/router/client.ts]: moved ✅

**Review by dai-shi — APPROVED** [review]
Let's see how this goes.

**Comment by wesbos** [comment]
Is there a way to use <Link> and not have the page scroll? NExt.js has `scroll={false}` which is nice to prevent scrolling on things like pagination.

**Comment by tylersayshi** [comment]
> Is there a way to use <Link> and not have the page scroll? NExt.js has `scroll={false}` which is nice to prevent scrolling on things like pagination. 

Not yet, I'll add a way! Thanks for the feedback :)

**Comment by dai-shi** [comment]
@wesbos Hi, thanks for the feedback.

We could consider adding such option.
Can you elaborate the use case to help us understand the limitation?

(I would like to keep `<Link>` as simple as `<a>`, and add flexibility to `useRouter`, because I believe `<a>` doesn't have `scroll={false}` property. But, the idea might change with your explanation.)

**Comment by wesbos** [comment]
Thanks for the reply - Here is a use case for where I need it.  Most frameworks provide this, and I fear pushing people to use a hook will result in them using buttons over links too often. 

https://github.com/user-attachments/assets/69d6ad57-5815-4fbe-8b15-320670ff3e7d

**Comment by tylersayshi** [comment]
> and I fear pushing people to use a hook will result in them using buttons over links too often.

completely agreed - thank you!

**Comment by dai-shi** [comment]
Yeah, that's convincing. Thanks for the explanation with the demo, Wes!

@tylersayshi Let's do that.

**Pull request merged**